### ISSUE TITLE
Fold similar issues

### ISSUE THREAD

**Issue by lexaknyazev** [KhronosGroup/glTF-Validator#29]
From https://github.com/KhronosGroup/glTF-Validator/issues/19#issuecomment-337367075
> It's just a case of potentially many errors being generated for a "tiny" mistake (i.e. a wrong `max` value). A single message like *"There is a value of 5.5 at index 45, which is greater than the maximum of 1.0, and there are 48343 more values exceeding the maximum"* could alleviate this.

**Comment by lexaknyazev** [comment]
Certain binary data issues should be folded when:
- `code` and `pointer` are the same, and
- number of issues exceeds some threshold (default 1?).

**Comment by lexaknyazev** [comment]
@emackey @javagl @donmccurdy 

Without changing current validation report schema, there's only one option on how to implement this: alter the content of `issues.messages[].message` field to be less specific.

**Current state**
```json
[
    {
        "code": "ACCESSOR_ELEMENT_OUT_OF_MAX_BOUND",
        "severity": 0,
        "pointer": "/accessors/2/max",
        "message": "Accessor element 59.6924934387207 at index 0 is greater than declared maximum value 6.352466583251953."
    },
    {
        "code": "ACCESSOR_ELEMENT_OUT_OF_MAX_BOUND",
        "severity": 0,
        "pointer": "/accessors/2/max",
        "message": "Accessor element 22.60923957824707 at index 21 is greater than declared maximum value 6.352466583251953."
    }
]
```

Other option could be to extend schema a bit:
```json
[
    {
        "code": "ACCESSOR_ELEMENT_OUT_OF_MAX_BOUND",
        "severity": 0,
        "pointer": "/accessors/2/max",
        "message": "2 values are exceeding the maximum.",
        "details": [
            "Accessor element 59.6924934387207 at index 0 is greater than declared maximum value 6.352466583251953.",
            "Accessor element 22.60923957824707 at index 21 is greater than declared maximum value 6.352466583251953."
        ]
    }
]
```

We could think of making it more machine-readable to generalize with #35.

What do you think?

**Comment by donmccurdy** [comment]
Curious to hear from others who would have implemented exporters, but personally I think it would be enough to just report what the actual maximum was, without reporting elements that violated the expected maximum. 

Another idea, somewhat related, would be to add a `maxIssuesPerType` option (in addition to `maxIssues` that would limit the number of issues reported of any particular type. When working o GLTFExporter, it has been generally enough to have only a few examples of the problem, rather than the complete list.

**Comment by lexaknyazev** [comment]
> I think it would be enough to just report what the actual maximum was, without reporting specific elements that violated the expected maximum.

I still regard those as two different issues.
1. Declared `accessor.max` value is _greater_ than the actual maximum. In this case, stored values don't violate anything. Reported once.
2. Declared `accessor.max` value is _less_ than the actual maximum. Reported for each value.

This issue is more general than just bounds - almost all binary data related checks could report per vertex.

> When working on GLTFExporter, it has been generally enough to have only a few examples of the problem, rather than the complete list.

That's why issues were grouped by `code` in the very first validation report schema. Going forward, I think the limit should apply per tuple `<code, pointer>` rather than per `code` only.

**Comment by donmccurdy** [comment]
In my mind, whether `max` is too large or too small it is still basically just wrong, and the steps I will take to debug the problem do not require every value. At most I might want a message like:

```
Expected maximum was `8`, but found larger elements at indices: 0, 4, 7, 10, 14, ...
```

^That list of indices can be truncated to N values, and that would be sufficient for me. But I certainly don't mind having everything reported under `details` if you're so inclined. 🙂

> That's why issues were grouped by code in the very first validation report schema. Going forward, I think the limit should apply per tuple <code, pointer> rather than per code only.

I'm still in favor of having a single flat list, it's just easier to work with in my own code, easier to represent in a protocol buffer, etc. As for `code` or a tuple — either way. It is also easy enough to add this limit in my viewer, I just haven't done that, so it's not a critical feature request. :)

**Comment by lexaknyazev** [comment]
> In my mind, whether `max` is too large or too small it is still basically just wrong, and the steps I will take to debug the problem do not require every value.

That's true when you produce a complete glTF file at once. I expect that tools/workflows which alter parts of existing glTF files may benefit from having a more detailed report.

> I'm still in favor of having a single flat list

So do I now. 

To cover #35, I'd like to have some generalized way of supplying homogeneous arrays of invalid data.

**Comment by emackey** [comment]
Early versions of the Substance Painter glTF exporter produced too-small `max` values for the input mesh (since fixed, I believe).  I still have the old buggy exporter, and when I get a model out of it I take these steps manually:

1. In VSCode, I scroll to the min/max section, which shows hundreds of overlapping errors.
2. Manually set all 3 `max` values to something too high, like 9999.
3. Now the validator stops complaining about hundreds of verts, and just tells me the correct max value to use for each axis in a single message. ("Supplied value 9999 does not match computed value 123.456" or some such).
4. I copy-and-paste the correct value out of the validation message and into the glTF.  Solved.

What I'm saying is this:  The vertex data is the source of truth.  If the verts don't agree with the min or max, it is *always* better to assume that min/max are wrong, not the verts.  If the individual verts were wrong the whole model would be the wrong shape, and that's not something we can produce a message for.

IMHO, there should be a single validation message, very similar to the one produced now when the max is too high, that the stated min or max value does not match the actual computed value (providing both values for comparison as it does now).  There is no need to call out individual verts that are "in violation" of the stated min or max.  The verts are the source of truth, just show the computed min/max compared to the provided one, and say the provided one is wrong.

**Comment by lexaknyazev** [comment]
Sorry, I don't follow how excessive verbosity of one particular message (which could be refined, ofc) affects a general case for reporting lists with wrong positions.

Here're all current validation messages besides min/max bounds that could be repeated many times for the same accessor:
- `ACCESSOR_ANIMATION_INPUT_NEGATIVE`
- `ACCESSOR_ANIMATION_INPUT_NON_INCREASING`
- `ACCESSOR_INDECOMPOSABLE_MATRIX`
- `ACCESSOR_INDEX_OOB`
- `ACCESSOR_INVALID_FLOAT`
- `ACCESSOR_INVALID_SIGN`
- `ACCESSOR_NON_UNIT`
- `ACCESSOR_SPARSE_INDEX_OOB`
- `ACCESSOR_SPARSE_INDICES_NON_INCREASING`

Are you saying that we should collapse each of them to just one message with truncated details embedded in plain text?

**Comment by emackey** [comment]
My own comments were specifically about min/max.  I'm fine with these others calling out individual violations.

Min/max is different because they are always computed from the vertex/accessor data.  The individual data elements are not computed using min/max space as bounds.

**Comment by emackey** [comment]
That said... Maybe the next major version of the validation schema could contain a list of indicies as a child of one message.  Maybe that would allow a lot of overlapping errors to be collapsed to a single error, but still result in a machine-readable list of where the problems are located?  Haven't thought much about this idea yet.

**Comment by lexaknyazev** [comment]
> If the individual verts were wrong the whole model would be the wrong shape, and that's not something we can produce a message for.

After thinking a bit about this, I admit that there's not much value in calling out thousands of verts with their individual values. However, the error message can say how many verts are in violation (in addition to stated and actual bound values). This would address my concerns.

**Comment by lexaknyazev** [comment]
Given that there is some interest in this feature, could we agree on the validation report schema updates?
The schema should be human- and machine-readable at the same time. 
This feature may be initially implemented behind a flag if that's needed.

**Comment by donmccurdy** [comment]
The issue seems specific to accessor values, so perhaps rather than a human-readable list of strings in `details`, we do something like this:

```
[
    {
        "code": "ACCESSOR_ELEMENT_OUT_OF_MAX_BOUND",
        "severity": 0,
        "pointer": "/accessors/2/max",
        "message": "2 values are exceeding the maximum.",
        "accessorIndices": [ 0, 21 ]
    }
]
```

... or ...

```
[
    {
        "code": "ACCESSOR_ELEMENT_OUT_OF_MAX_BOUND",
        "severity": 0,
        "pointer": "/accessors/2/max",
        "message": "2 values are exceeding the maximum.",
        "accessorEntries": [ { index: 0, value: 59.2800 }, { index: 28, value: 22.3208 } ]
    }
]
```

The only reason I can think of to include the value is that sometimes the reported issue is quite minor, like normals with a length of `1.00047`. If I could work around that with a `precision` argument to the validator, I probably wouldn't care whether accessor values were reported. It is also unclear would a "value" would be for e.g. ACCESSOR_INDECOMPOSABLE_MATRIX.

**Comment by lexaknyazev** [comment]
> It is also unclear what a "value" would be for e.g. ACCESSOR_INDECOMPOSABLE_MATRIX.

`value` could be an optional field, however I'm in favor of providing it when it's possible (not sure about matrices). 

I think that two flat arrays would be a better option (performance-wise):
```json
[
    {
        "code": "ACCESSOR_ELEMENT_OUT_OF_MAX_BOUND",
        "severity": 0,
        "pointer": "/accessors/2/max",
        "message": "2 values are exceeding the maximum.",
        "accessorIndices": [ 0, 28 ],
        "accessorValues": [ 59.2800, 22.3208 ]
    }
]
```

We have a separate issue for precision-related issues. I'd prefer the spec having defined thresholds for them.

By "human-readable" I mean that the message should still contain meaningful summary, while low-level details could be left to tools.
